feat: social tips contract

manu0466 opened the pull request:
## Description
Implementation of the [ADR-005](https://github.com/desmos-labs/desmos-contracts/blob/master/docs/architecture/adr-005-social-tips-contract.md).

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos-contracts/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos-contracts/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed contract logic
- [ ] reviewed contract security
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

manu0466 commented:
@dadamu @bragaz During the developement i noticed that the current specification don't take in account that a bad actor can abuse of the pending tips to store unlimited tips in the contract state. 
Since there is this problem I have: 
1. Added a parameter in the `InstantiateMsg` to specify the max allowed pending tips that the contract can store for each user and capped this value to a maximum of 20;
2. Added a `UpdateMaxPendingTips` message that can be called from the admin to update the maximum allowed pending tips for each user.

Since this is change from the ADR I ask a feedback from you guys

dadamu commented:
@manu0466 I agree with you. Additionally, we should allow senders to cancel the pending tip since the recipient on web3 may not join the Desmos, it would occupy a slot forever.

manu0466 commented:
@dadamu @bragaz I have updated the contract with:
1. Allow the user to remove the a previously sent tip that hasn't be clamed
2. Enforced a limit for the amount of pending tips that a single user can create, this is to pretect against the case of an user that creates a big amount of tips that are pending to different app link
3. Enforced that an user can send only a single tip to central application handle that can not be sent directly to the final user
4. Added `UnclaimedSentTips` query message to know the tips that an user has sent that are currently holded from the contract
5.  Query message to quert the contract's configurations
All this changes are also inside the ADR document.

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
````diff
@@ -31,6 +32,19 @@ Here below the specifications for the contract's messages:
+}
+```
+
+* The `admin` identifies the user that controls the contract, if is `None` the contract will pick the address of who sent the transaction;
````
```suggestion
* The `admin` identifies the user that controls the contract, if it is `None` the contract will pick the address of who sent the transaction;
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
```diff
@@ -49,14 +67,33 @@ The `MessageInfo` fields contains:
 * the funds necessary to cover the fees plus the tip amount specified inside the `funds` field
 * the sender identified by the `sender` field
 
+**NOTE** If the user have already sent a tip to the same centralized application reference the contracts collects
```
```suggestion
**NOTE** If the user has already sent a tip to the same centralized application reference the contracts collects
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
```diff
@@ -49,14 +67,33 @@ The `MessageInfo` fields contains:
 * the sender identified by the `sender` field
 
+**NOTE** If the user have already sent a tip to the same centralized application reference the contracts collects
+only the last one sending back to the user the previous tip amount.
```
```suggestion
only the last one then sending back to the user the previous tip amount.
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
```diff
@@ -49,14 +67,33 @@ The `MessageInfo` fields contains:
 owns the identities to which the tips refer.
 
+##### UpdateAdmin
+With the `UpdateAdmin` message the current admin can update the contract's admin.
```
```suggestion
With the `UpdateAdmin` message, the current admin can update the contract's admin.
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
```diff
@@ -49,14 +67,33 @@ The `MessageInfo` fields contains:
+With the `UpdateAdmin` message the current admin can update the contract's admin.
+
+### UpdateMaxPendingTips
+With the `UpdateMaxPendingTips` message the current admin can change the amount of pending tips that
```
```suggestion
With the `UpdateMaxPendingTips` message, the current admin can change the amount of pending tips that
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
```diff
@@ -49,14 +67,33 @@ The `MessageInfo` fields contains:
+can be associated to a centralized application user.
+
+### UpdateMaxSentPendingTips
+With the `UpdateMaxPendingTips` message the current admin can change the amount of tips that
```
```suggestion
With the `UpdateMaxPendingTips` message, the current admin can change the amount of tips that
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
```diff
@@ -49,14 +67,33 @@ The `MessageInfo` fields contains:
+the contract collects for a specific sender address.
+
+### RemovePendingTip
+With the `RemovePendingTip` message the user can cancel a tip that haven't been claimed yet.
```
```suggestion
With the `RemovePendingTip` message, the user can cancel a tip that haven't been claimed yet.
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
````diff
@@ -74,3 +111,29 @@ struct Tip {
 ```
+
+##### UnclaimedSentTips
+With the `UnclaimedSentTips` message the user can query the tips that has sent that aren't claimed yet.
````
```suggestion
With the `UnclaimedSentTips` message, the user can query the unclaimed tips that has been sent.
```

dadamu commented on docs/architecture/adr-005-social-tips-contract.md:
````diff
@@ -74,3 +111,29 @@ struct Tip {
+```
+
+### Config
+With the  `Config{}` message a user can query the current contract's configurations.
````
```suggestion
With the  `Config{}` message, a user can query the current contract's configurations.
```

dadamu commented on contracts/social-tips/rustfmt.toml:
```diff
@@ -0,0 +1,15 @@
+# stable
```
We can remove it since there is a general fmt config outside.

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,142 @@
+
+#[cw_serde]
+pub enum ExecuteMsg {
+    /// Message to send a tip to another user.
```
```suggestion
    /// Message to send a tip to another user by the application handle.
```

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,142 @@
+pub enum ExecuteMsg {
+    /// Message to send a tip to another user.
+    SendTip { application: String, handle: String },
+    /// Message that allow a user to claim their tips.
```
```suggestion
    /// Message that allows a user to claim their tips.
```

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,142 @@
+    /// Message that allows the current admin to update the contract admin.
+    UpdateAdmin { new_admin: String },
+    /// Message that allows the current admin to update the max pending tips that
+    /// can be associated to a centralize application.
```
```suggestion
    /// can be associated to a centralized application.
```

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,142 @@
+#[cw_serde]
+#[derive(QueryResponses)]
+pub enum QueryMsg {
+    /// Query the pending tips of an user.
```
```suggestion
    /// Query the pending tips of a user.
```

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,142 @@
+    /// Query the pending tips of an user.
+    #[returns(QueryPendingTipsResponse)]
+    UserPendingTips { user: String },
+    /// Message to query the unclaimed tips sent from an user.
```
```suggestion
    /// Message to query the unclaimed tips sent from a user.
```

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+
+#[cfg(test)]
+mod tests {
+    use crate::contract::{execute, instantiate, query};
```
We can use `use super::*` to import all the dependancies as this file.

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+    let new_admin_addr = deps.api.addr_validate(&new_admin)?;
+
+    CONFIG.update(deps.storage, |mut config| {
+        if config.admin != new_admin_addr {
```
It should check if sender is admin, not new admin address.

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+    let sender = info.sender;
+    let funds = sum_coins_sorted(info.funds)?;
+
+    if funds.is_empty() {
```
Empty check can be before `sum_coins` to return error without unnecessary summing.

manu0466 replied:
`sum_coins_sorted` don't do any sum if the length of the provided `Vec` is <= 1.
I do prefer keep the check ther so that the check is perfomed on the output of `sum_coins_sorted` since can be different from `info.funds`.

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+        None,
+    )?;
+
+    // Fail if there are more then 1 users with the same application link.
+    if response.owners.len() > 1 {
+        return Err(ContractError::ToManyOwners {});
+    }
```
The check can be removed since all address linked to the application handle is considered to have right to claim the tips.

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+
+    let merged_coins = sum_coins_sorted(coins)?;
+
+    if merged_coins.is_empty() {
+        return Err(ContractError::NoTipsAvailable {
+            user: info.sender.to_string(),
+        });
+    }
```
It can be moved before `sum_coins_sorted` to return error without summing.

manu0466 replied:
as above

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+    info: MessageInfo,
+    value: u16,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+
+    if config.admin != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
```
We can have a function to check if the sender is admin or not. Then it can be replaced to something like:
```
check_admin(&deps.storage, sender)?;
```

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+        None,
+    )?;
+
+    // Fail if there are more then 1 users with the same application link.
+    if response.owners.len() > 1 {
+        return Err(ContractError::ToManyOwners {});
+    }
```
Should we remove it? From my side, all addresses linked to the application should be considered the valid to send the tips. What do you think?

manu0466 replied:
Ok, I can add a `owner_index: Option<Uint64>` to `SendTip` so that the user can select to witch addres the tip will be sent and if is `None` I will pick the first owner item.

dadamu commented on contracts/social-tips/src/utils.rs:
```diff
@@ -0,0 +1,49 @@
+/// Iterates over the coins vector and merges the coins having the same `denom`
+/// and return them sorted by denom.
+/// * `coins` - Vector of coins to merge.
+pub fn sum_coins_sorted(coins: Vec<Coin>) -> Result<Vec<Coin>, ContractError> {
```
Missing tests for it.

dadamu commented on contracts/social-tips/src/utils.rs:
```diff
@@ -0,0 +1,49 @@
+
+/// Serialize a slice of [`Coin`] into where each coin is separated by a "," (comma).
+/// * `coins` - Coins slice to serialize.
+pub fn serialize_coins(coins: &[Coin]) -> String {
```
As above.

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,142 @@
+}
+
+impl ExecuteMsg {
+    pub fn validate(&self) -> Result<(), ContractError> {
```
Can we move `msg.validate()?` tests here from `contract.rs`. From my side, Message to pass to `execute` should be considered as valid.

manu0466 replied:
I will add the tests, i do prefer having also test in `contract.rs` with invalid messages to make sure that `msg.validate` has been called from the contract.

dadamu replied:
I think we can remain only one to `*_with_invalid_msg_error` for each execution msgs. What do you think?

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1346 @@
+// consts
+const ATTRIBUTE_ACTION: &str = "action";
+const ATTRIBUTE_TIP_COLLECTED: &str = "tip_collected";
+const ATTRIBUTE_TIP_RECEIVER: &str = "tip_receiver";
```
I would name receiver to claimer.

dadamu commented on contracts/social-tips/schema/execute_msg.json:
```diff
@@ -0,0 +1,39 @@
+{
```
The schem json is needed to be updated.

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1375 @@
+            .add_attribute(ATTRIBUTE_TIP_AMOUNT, serialized_coins);
+
+        // Send back the funds of the replaced tip.
+        if let Some(replaced_tip) = replaced {
+            response = response.add_message(BankMsg::Send {
+                amount: replaced_tip.amount,
+                to_address: replaced_tip.sender.to_string(),
+            });
+        }
+
+        Ok(response)
+    };
```
Since it is related to `x/bank`, I would have a integration test to test the final balance of the sender.

manu0466 replied:
There are already unit tests to ensure that the proper message has ben added to the response, I think is more than enough to ensure the correct behaviour of the contract.

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1375 @@
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_CLAIM_PENDING_TIPS)
+        .add_attribute(ATTRIBUTE_TIP_CLAIMER, &info.sender)
+        .add_attribute(ATTRIBUTE_TIP_AMOUNT, serialize_coins(&merged_coins))
+        .add_message(BankMsg::Send {
+            amount: merged_coins,
+            to_address: info.sender.to_string(),
+        }))
```
As above.

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1375 @@
+                ATTRIBUTE_REMOVED_TIP_AMOUNT,
+                serialize_coins(&to_remove_tip.amount),
+            )
+            .add_message(BankMsg::Send {
+                amount: to_remove_tip.amount,
+                to_address: refund_address,
+            }))
```
As above.

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,344 @@
+        match self {
+            ExecuteMsg::SendTip {
+                application,
+                handle: handler,
```
I would use the name `handle`, instead.

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,344 @@
+                }
+
+                if handler.is_empty() {
+                    return Err(ContractError::InvalidUserHandler {});
```
The error should be `InvalidUserHandle` from my side.

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,344 @@
+            }
+            ExecuteMsg::RemovePendingTip {
+                application,
+                handle: handler,
```
As above, I would use the name handle.

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,344 @@
+                }
+
+                if handler.is_empty() {
+                    return Err(ContractError::InvalidUserHandler {});
```
As above, I would name the error `InvalidUserHandle`

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1375 @@
+    use desmos_bindings::query::DesmosQuery;
+
+    const ADMIN: &str = "admin";
+    const USER_1: &str = "user1";
+    const USER_2: &str = "user2";
+
```
I would name them to `SENDER` and `CLAIMER` to be exiplicit on usecases.

dadamu commented on contracts/social-tips/Cargo.toml:
```diff
@@ -0,0 +1,56 @@
+optimize = """docker run --rm -v "$(pwd)":/code \
+  --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target \
+  --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
+  cosmwasm/rust-optimizer:0.12.6
```
cargo optimize is not feasible since it should be `workspace-optimizer`, instead.

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    #[test]
+    fn update_max_pending_tips_with_not_bigger_value_error() {
```
```suggestion
    fn update_max_pending_tips_with_bigger_max_pending_value_error() {
```

dadamu commented on contracts/social-tips/src/msg.rs:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    #[test]
+    fn update_max_sent_pending_tips_with_not_bigger_value_error() {
```
```suggestion
    fn update_max_sent_pending_tips_with_bigger_max_sent_pending_value_error() {
```

dadamu commented on contracts/social-tips/src/contract.rs:
```diff
@@ -0,0 +1,1375 @@
+    }
+
+    #[test]
+    fn remove_not_existing_pending_tip_error() {
```
```suggestion
    fn remove_non_existing_pending_tip_error() {
```

dadamu commented:
Overall looks good to me, just some minor issues about test naming.

dadamu commented on contracts/social-tips/examples/schema.rs:
```diff
@@ -0,0 +1,22 @@
+    create_dir_all(&out_dir).unwrap();
+    remove_schemas(&out_dir).unwrap();
+
+    export_schema(&schema_for!(InstantiateMsg), &out_dir);
+    export_schema(&schema_for!(ExecuteMsg), &out_dir);
+    export_schema(&schema_for!(QueryMsg), &out_dir);
+    export_schema(&schema_for!(QueryPendingTipsResponse), &out_dir);
+    export_schema(&schema_for!(QueryUnclaimedSentTipsResponse), &out_dir);
+    export_schema(&schema_for!(QueryConfigResponse), &out_dir);
```
The reason why you change `write_api` to `export_schema` is because the `ts-codegen` does not work properly, right?
Maybe we should try [new ts codegen](https://github.com/CosmWasm/ts-codegen) built by Cosmwasm team.

dadamu replied:
Should we revert the changes of other contract schemas to `export_schema` as well before updating ts-codegen?

manu0466 replied:
Yes, I have reverted the changes for that reason. Yes the changes should be reverted to use the new `write_api`.
